feat(api): Endpoints for classroom - exchanging user data

NewtonLC opened the pull request:
Co-Authored by:
@CarlyAThomas
@ZhenyuYu1

Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or Gitpod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes https://github.com/freeCodeCamp/classroom/issues/548


This issue calls for the Classroom App to retrieve real users' data directly from freeCodeCamp proper, which is vital for making the Classroom App work. This will allow teachers to monitor students' progress and activity in the freeCodeCamp curriculum.

<!-- Feel free to add any additional description of changes below this line -->

This PR creates an API endpoint in freeCodeCamp that allows the Classroom App to fetch students' completed challenge data from their usernames. Since users' completedChallenges attribute does not contain metadata about the completed challenges(name, certification, etc.), we also created an endpoint that maps challenge IDs to challenge metadata, then returns this map. The Classroom App will call this endpoint in order to retrieve the map, which will be used to help display the challenges.

Commit:
feat: created new classroom API endpoints for exchanging user data
Co-Authored by:
Carly Thomas <CarlyAThomas@users.noreply.github.com>
Zhenyu Yu <ZhenyuYu1@users.noreply.github.com>

Both endpoints are to be called from the Classroom app.
- Added new endpoint to contruct a map of challenge IDs to name, certification, etc. based on _meta files, then return this map.
- Added new endpoint to take in a list of student emails, then return an array of those users' completed challenge data.

raisedadead reviewed (changes requested):
Action Items:

@raisedadead: 

- will create the curriculum server repository with all the wiring in it so the @freeCodeCamp/classroom team can expand the same as per the needs of the classroom app
- will document steps for the Classroom team to setup and start working on it. It will be available in the project README.

@freecodecamp/classroom:

- needs to remove the build curriculum endpoints entirely, since the functionality will be made available via the above.

This PR:

- will need another endpoint to send a single email and retrieve the user id (camper's id), classroom can store the User id in its own DB. The user id will not change (emails can).
- the get-user-data endpoint needs to:
  - send user ids (batch is OK) to retrieve the data it needs
  - check the user consent (more on this later) is given before sending the data out
  
User consent:

- will need a separate PR (@raisedadead will share code examples) to add a check box to the settings page UI
- the check box can map to a `isClassroomAccount` prop on the user -- false by default

raisedadead commented:
I have left my quick notes on this PR from the meeting. We can discuss more on this async if you prefer in the Google Chat room we have for this.

NewtonLC committed:
Merge branch 'freeCodeCamp:main' into challenge-map-newton

NewtonLC committed:
Removed Challenge Map endpoint and Updated Classroom Endpoints

Co-authored-by: Carly Thomas <CarlyAThomas@users.noreply.github.com>

raisedadead commented:
Great! https://github.com/freeCodeCamp/curriculum-db is also coming along nicely.

ojeytonwilliams commented on api/src/routes/protected/classroom.ts:
```diff
@@ -0,0 +1,147 @@
+import { FastifyPluginCallback } from 'fastify';
+import { Type } from '@fastify/type-provider-typebox';
```
```suggestion
import { FastifyPluginCallbackTypebox } from '@fastify/type-provider-typebox';
```
We need this type because it uses the `schema` to generate types for the request and reply objects.

ojeytonwilliams commented on api/src/routes/protected/classroom.ts:
```diff
@@ -0,0 +1,147 @@
+  fastify.post(
+    '/api/protected/classroom/get-user-data',
+    {
+      schema: {
```
Please move this to the same folder as the other schema.

ojeytonwilliams commented on api/src/routes/protected/classroom.ts:
```diff
@@ -0,0 +1,147 @@
+                  id: Type.String(),
+                  completedDate: Type.Number(),
+                  challengeName: Type.Optional(Type.String()),
+                  files: Type.Optional(Type.Array(Type.Object({}))),
```
If you're using the files in classroom, then I recommend specifying the shape of the object, so you can be sure that the data is what you expect.

NewtonLC replied:
Hi Oliver! Upon further review, it seems that we don't actually use files in classroom, so I will go ahead and remove this line from the PR.

ojeytonwilliams commented on api/src/routes/protected/classroom.ts:
```diff
@@ -0,0 +1,147 @@
+      }
+    },
+    async (request, reply) => {
+      const { userIds = [] } = request.body as { userIds: string[] };
```
```suggestion
      const { userIds = [] } = request.body
```
Also not necessary when using a schema.

ojeytonwilliams commented on api/src/routes/protected/classroom.ts:
```diff
@@ -0,0 +1,147 @@
+        });
+
+        // Map to transform user data into the required format
+        const userData: Record<string, object[]> = {};
```
```suggestion
        const userData: Record<string, NormalizedChallenge[]> = {};
```
object is a bit vague. I would specify precisely what you expect

ojeytonwilliams commented on api/src/routes/protected/classroom.ts:
```diff
@@ -0,0 +1,147 @@
+          select: {
+            id: true,
+            isClassroomAccount: true
+          }
```
```suggestion
          where: { email, isClassroomAccount: true },
          select: { id: true }
```
i.e. select only what you're interested in, rather than getting everything and filtering.

ojeytonwilliams reviewed (changes requested):
Thanks for the updates @NewtonLC and @CarlyAThomas, this looks like what we want. I found a few things, but nothing serious - just minor tweaks and suggestions to be a little more consistent with the rest of the codebase.

NewtonLC committed:
Implement Code Review Suggestions

Co-authored by: Carly Thomas <CarlyAThomas@users.noreply.github.com>

CarlyAThomas commented:
We have updated the PR based on your suggestions. One note though, we are still looking into how we would like to specify the shape of the userData object and will need to get back to you on this specific.

Thank you for reviewing our PR!

NewtonLC committed:
Removed Unused Files and Information from get-user-data Endpoint

Co-authored-by: Carly Thomas <CarlyAThomas@users.noreply.github.com>

ojeytonwilliams reviewed (changes requested):
Sorry for the lack of updates, but this approach should be fine.

Could you add some simple tests to lock this down? Something that checks the expected output would have caught the fact that `challengeName` doesn't exist.

NewtonLC committed:
Update api/src/schemas/classroom/classroom.ts

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

NewtonLC committed:
Added Test File for Classroom Endpoints
Additionally, edited the  body to not require
a max of 50 items, so a custom error message will be used.

NewtonLC committed:
Added Test File for Classroom Endpoints
Additionally, edited the  body to not require
a max of 50 items, so a custom error message will be used.

NewtonLC commented:
Hello Oliver! I added some simple test cases for the classroom endpoints. Here's a quick summary:
`get-user-ids` tests:
Call with incorrectly formatted email --> 400 Error
No classroom account matches email --> 404 Error
A classroom account matches email --> 200 Status, return camper ID
Other failed database query --> 500 Error

`get-user-data` tests:
Call with more than 50 camper IDs --> 400 Error
Call with 50 or fewer camper IDs --> 200 Status, return record of camper's data
Other failed database query --> 500 Error

Unauthenticated call --> 401 Error

NewtonLC committed:
Removed fastify error from response schema

NewtonLC committed:
Updated get-user-id User Not Found response
Changed 404 error to 200 response, along with empty student ID

NewtonLC committed:
Re: Updated get-user-id User Not Found response
Also edited to give 200 responses for invalid email formats

NewtonLC commented:
Hello! Just a quick note to say that this PR is ready for review. Thank you!

utsab commented:
Hi @ojeytonwilliams and @raisedadead -- I wanted send a quick follow up about this PR.  I believe @NewtonLC and @CarlyAThomas addressed all of your feedback for the PR including writing the test cases.  Would you be able to take another look at the PR and let us know if it's ready to be merged?  Thank you.
